### PR TITLE
feat: Automatically calculate the current version number through Git…

### LINKED ISSUE #537: Governance: Automatically calculate the current version number through Git
### What would you like to be added?

Automatically calculate the current version number through Git.

For example, assuming the latest tag is `v0.4.2`, if `HEAD` is equal to `v0.4.2`, then the version number is `v0.4.2`. If not equal, the version number is `v0.4.3-xxxxxx`, among them, `xxxxxx` is `${shortSha}`.

### Why is this needed?

The current version number is hardcode in the [pkg/version/VERSION](https://github.com/KusionStack/karpor/blob/main/pkg/version/VERSION) file, and its value needs to be manually modified before each new version is released, which is not conducive to long-term maintenance.

### PR DESCRIPTION
**Pull request by rajeshkio**
## What type of PR is this?
/kind feature

## What this PR does / why we need it:

 Automates versioning and updates pkg/version/VERSION file

## Which issue(s) this PR fixes:

Fixes #537

### CODE REVIEW

**Comment by rajeshkio** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by rajeshkio** [comment]
recheck

File: pkg/version/scripts/gen/gen.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	}
+	tagCommit, err := gitutil.GetTagCommit(latestTag)
+	if err != nil {
+		fmt.Println("Error getting tag commit:", err)
+		return "", err
+	}
+	if currentHead == tagCommit {
+		return latestTag, nil
+	}
+	return fmt.Sprintf("%s-g%s", latestTag, currentCommit), nil
```
Comment (elliotxx) [inline pkg/version/scripts/gen/gen.go:40]: Is there a typo here? `s/%s-g%s/%s-%s`?

**Comment by SparkYuan** [comment]
@rajeshkio Welcome to join our community. Here is our [slack](https://join.slack.com/t/kusionstack/shared_invite/zt-2drafxksz-VzCZZwlraHP4xpPeh_g8lg)

**Comment by elliotxx** [comment]
@rajeshkio  Nice! It seems to be able to automatically calculate the current version number, but, the make command still needs to be manually executed before each release. So I am thinking about how to integrate it in the release pipeline. Maybe we need to add a hook to the .goreleaser.yml. Do you have any good ideas?

**Comment by rajeshkio** [comment]
sure. Let me see.

**Comment by elliotxx** [comment]
@rajeshkio You can refer to this configuration: https://github.com/elliotxx/go-web-template/blob/master/.goreleaser.yml#L7

But this can only ensure that the version in the Image is correct and cannot guarantee that `go Install` can get the correct version number.

**Comment by rajeshkio** [comment]
make sense to hook it just after go mod tidy?

`before:
  hooks:
    # You may remove this if you don't use go modules.
    - go mod tidy
    - go run pkg/version/scripts/gen/gen.go`

**Comment by rajeshkio** [comment]
> @rajeshkio You can refer to this configuration: https://github.com/elliotxx/go-web-template/blob/master/.goreleaser.yml#L7
> 
> But this can only ensure that the version in the Image is correct and cannot guarantee that `go install` can get the correct version number.

This is interesting. Something I have never experienced. Can I test it locally?

**Comment by elliotxx** [comment]
> > @rajeshkio You can refer to this configuration: https://github.com/elliotxx/go-web-template/blob/master/.goreleaser.yml#L7
> > But this can only ensure that the version in the Image is correct and cannot guarantee that `go install` can get the correct version number.
> 
> This is interesting. Something I have never experienced. Can I test it locally?

Of course, this is some ways to test this case locally:
- Dry Run Mode for Goreleaser: https://goreleaser.com/quick-start/#dry-run
- `go install github.com/rajeshkio/karpor/cmd@automatically-calculate-current-version`

**Comment by elliotxx** [comment]
> make sense to hook it just after go mod tidy?
> 
> before: hooks: # You may remove this if you don't use go modules. - go mod tidy - go run pkg/version/scripts/gen/gen.go

Yes, this hook ensures that the `gen` script is executed to update the version number before the `go build`.

**Comment by elliotxx** [comment]
Maybe we do n’t need to consider `go install` for the time being. This is just considered for the future `Standalone` installation mode (such as `homebrew`, `scoop`, `go install`).

**Comment by rajeshkio** [comment]
> Maybe we do n’t need to consider `go install` for the time being. This is just considered for the future `Standalone` installation mode (such as `homebrew`, `scoop`, `go install`).

okay. Your decision. I am not well aware of this. If it can be ignored for now, we can create another issue for this, assigned to me, and I can work on it separately.

**Comment by rajeshkio** [comment]
> > make sense to hook it just after go mod tidy?
> > before: hooks: # You may remove this if you don't use go modules. - go mod tidy - go run pkg/version/scripts/gen/gen.go
> 
> Yes, this hook ensures that the `gen` script is executed to update the version number before the `go build`.

added.

**Comment by rajeshkio** [comment]
@elliotxx Let me know if anything is pending from my side

**Comment by elliotxx** [comment]
@rajeshkio You have done a good job, I just have no time to do full test. But I thought about it, this it doesn't matter. We are the trunk-based development mode, there will be a full test before the official version is released.

**Review by elliotxx — APPROVED** [review]
lgtm

**Comment by elliotxx** [comment]
@rajeshkio Welcome to become a project contributor! #552  Looking forward to more communication between us 🍻

**Pull request merged**